### ISSUE TITLE
Need a tool to extract actor or class signatures

### ISSUE THREAD

**Issue by vporton** [dfinity/motoko#4753]
We need to have a tool that will extract actor/class signatures to create "header" files. I will explain on an example:

```motoko
actor class A() {
  public shared func f() {};
}
```
would be transformed into:
```motoko
module A {
  type A = actor {
    f: shared () -> ();
  };
}
```

**Comment by rvanasa** [comment]
Thank you for the suggestion! Do you have a motivating project / use case so we can better understand the problem you hope to solve with this tool?

**Comment by vporton** [comment]
> Thank you for the suggestion! Is there a motivating project / use case so we can better understand the problem you hope to solve with this tool?

Of course: I have a CanDB partition actor X dependent on actor Y. But Y needs to call methods from X. I can't make Y dependent on X, because it would be a circular dependency (and this also would greatly increase compilation time). So I would preprocess X to have an abstract actor to include into Y.

Currently I edit this abstract actor manually what is error-prone.

**Comment by rvanasa** [comment]
If I understand correctly, we currently have a way to do this with Motoko bindings from [`dfx generate`](https://internetcomputer.org/docs/current/developer-docs/developer-tools/cli-tools/cli-reference/dfx-generate). Here is an [example dfx configuration](https://github.com/internet-computer-protocol/evm-rpc-canister/blob/ed3447f5a11fb2722532ccc8d21b482cd2299fb1/dfx.json#L8-L11) which illustrates how this works in practice. 

If you want a fully custom solution, you could look into using [`file.parseMotokoTyped()`](https://github.com/dfinity/node-motoko?tab=readme-ov-file#file-api) from the [`motoko`](https://npmjs.com/package/motoko) npm package to process the AST of the Motoko actor in any way that you'd like.

**Comment by vporton** [comment]
This does not work in practice:

The following is a simplified idea of mutual dependence: in one direction dependency on the implementation and in other one just on the interface.

[Consider](https://github.com/vporton/bug-mo-interdep) `A.mo` depending on `B.mo` in implementation and `B.mo` dependent on `A.mo` as an interface:

`dfx.json`:
```json
{
  "canisters": {
    "A": {
      "main": "src/interdep_backend/A.mo",
      "type": "motoko",
      "dependencies": ["B"],
      "declarations": {
        "bindings": ["mo"]
      }
    },
    "B": {
      "main": "src/interdep_backend/B.mo",
      "type": "motoko"
    }
  },
  "defaults": {
    "build": {
      "args": "",
      "packtool": ""
    }
  },
  "output_env_file": ".env",
  "version": 1
}
```

`A.mo`:
```motoko
import B "canister:B";

actor A {};
```

`B.mo`:
```
import A "../declarations/A/A";

actor B {};
```

This is impossible to compile:

To compile `B.mo`, need first create `declarations/A/A.mo`:

```
$ dfx generate A
...
/home/porton/Projects/bug/interdep/src/interdep_backend/B.mo:1.1-1.31: import error [M0009], file "/home/porton/Projects/bug/interdep/src/declarations/A/A/lib.mo" does not exist
```

That is, generation of `A` depends on itself.

Any ideas how to fix it?

**Comment by rvanasa** [comment]
I asked about this internally, and it seems like there isn't really a way around this given the design of dfx. The cleanest solution would be to remove the mutual dependency if possible. I'll bring this up in the next Languages team meeting so we can decide on a priority for addressing this, perhaps with a header module code generation tool as you suggested.

**Comment by crusso** [comment]
Not that I'm recommending this approach but I did actually come up with a workaround for this using hand-crafted candid for the  forward declarations.

https://github.com/crusso/circ
